store.go

```diff
@@ -31,6 +32,100 @@ type Store interface {
+type StoreOptions struct {
+	// AllowPlaintextPut allows saving credentials in plaintext in the config
+	// file.
+	//   - If AllowPlaintextPut is set to false (default value), Put() will
+	//     return an error when native store is not available.
+	//   - If AllowPlaintextPut is set to true, Put() will save credentials in
+	//     plaintext in the config file when native store is not available.
+	AllowPlaintextPut bool
+}
+
+// NewStore returns a store based on given config file.
+// If no any authentication is configured in the config file, a platform-default
+// native store will be used.
+//   - Windows: "wincred"
+//   - Linux: "pass" or "secretservice"
+//   - macOS: "osxkeychain"
+//
+// Reference: https://docs.docker.com/engine/reference/commandline/login/#credentials-store
+func NewStore(configPath string, opts StoreOptions) (Store, error) {
+	cfg, err := config.LoadConfigFile(configPath)
+	if err != nil {
+		return nil, err
+	}
+	if !cfg.IsAuthConfigured() {
+		if defaultCredsStore := getDefaultHelperSuffix(); defaultCredsStore != "" {
+			cfg.CredentialsStore = defaultCredsStore
+		}
+	}
+
+	return &dynamicStore{
+		config:  cfg,
+		options: opts,
+	}, nil
+}
+
+// Get retrieves credentials from the store for the given server address.
+func (ds *dynamicStore) Get(ctx context.Context, serverAddress string) (auth.Credential, error) {
+	store, err := ds.getStore(serverAddress)
+	if err != nil {
+		return auth.EmptyCredential, err
+	}
+	return store.Get(ctx, serverAddress)
+}
+
+// Put saves credentials into the store for the given server address.
+// Returns ErrPlaintextPutDisabled if native store is not available and
+// StoreOptions.AllowPlaintextPut is set to false.
+func (ds *dynamicStore) Put(ctx context.Context, serverAddress string, cred auth.Credential) error {
+	store, err := ds.getStore(serverAddress)
+	if err != nil {
+		return err
+	}
+	return store.Put(ctx, serverAddress, cred)
+}
+
+// Delete removes credentials from the store for the given server address.
+func (ds *dynamicStore) Delete(ctx context.Context, serverAddress string) error {
+	store, err := ds.getStore(serverAddress)
+	if err != nil {
+		return err
+	}
+	return store.Delete(ctx, serverAddress)
+}
+
+// getHelperSuffix returns the credential helper suffix for the given server
+// address.
+func (ds *dynamicStore) getHelperSuffix(serverAddress string) string {
+	// 1. Look for a server-specific credential helper first
+	if helper := ds.config.GetCredentialHelper(serverAddress); helper != "" {
+		return helper
+	}
+	// 2. Then look for the configured native store
+	return ds.config.CredentialsStore
+}
+
+// getStore returns a store for the given server address.
+func (ds *dynamicStore) getStore(serverAddress string) (Store, error) {
+	if helper := ds.getHelperSuffix(serverAddress); helper != "" {
+		return NewNativeStore(helper), nil
+	}
+
+	fs := newFileStore(ds.config)
+	fs.DisablePut = !ds.options.AllowPlaintextPut
```

Comment:
I'm trying to understand the design logic here, so please correct me if I'm wrong.
Basically, we have 3 ways to tell if authentication is set up in a docker config file, they are:
1. credential helper
2. credential store
3. auths cache

This is a sort of waterfall logic, in other words, following the order of the list above. I guess my first question is: do we have a spec that explicitly decides which one goes first? If not, how did we decide this order again?
Another question: I'm not very sure why do we need the `File Store` struct? Looks like a `Native store` handles `credential helper` and `credential store`, and a `File Store` handles `auths cache`. Is there any concern that we handle them all in one place?

Reply:
> Do we have a spec that explicitly decides which one goes first? If not, how did we decide this order again?

`oras-credentials-go` tries to mimic the behavior of docker CLI. The references are the  
[Docker documentation](https://docs.docker.com/engine/reference/commandline/login/#default-behavior) and the [Docker code](https://github.com/docker/cli/blob/v24.0.0-beta.2/cli/config/configfile/file.go#L250-L257).

> I'm not very sure why do we need the File Store struct? Looks like a Native store handles credential helper and credential store, and a File Store handles auths cache. Is there any concern that we handle them all in one place?

It's just to make the code clean and easy to maintain.

Reply:
> I've updated the comment to call out the logic.

Cool. The comment is very helpful now. Thanks~